pkg/controllers/failure_policy.go

```diff
@@ -0,0 +1,174 @@
+		}
+	}
+
+	// We get here when no rule matched any of the failure policy rules
+	fmt.Println("never found a matching rule and returning nil to represent the default rule.")
+	return nil, buckets[defaultRuleIndex].firstFailedJob
+}
+
+// The type failurePolicyActionApplier applies a FailurePolicyAction and returns nil if the FailurePolicyAction was successfully applied.
+// The function returns an error otherwise.
+type failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error
+
+// failJobSetActionApplier applies the FailJobSet FailurePolicyAction
+var failJobSetActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
+	failureMessage := messageWithFirstFailedJob(constants.ReachedMaxRestartsMessage, matchingFailedJob.Name)
+	setJobSetFailedCondition(ctx, js, constants.ReachedMaxRestartsReason, failureMessage, updateStatusOpts)
+	return nil
+}
+
+// restartJobSetActionApplier applies the RestartJobSet FailurePolicyAction
+var restartJobSetActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
+	if js.Status.RestartsCountTowardsMax >= js.Spec.FailurePolicy.MaxRestarts {
+		failureMessage := messageWithFirstFailedJob(constants.ReachedMaxRestartsMessage, matchingFailedJob.Name)
+		setJobSetFailedCondition(ctx, js, constants.ReachedMaxRestartsReason, failureMessage, updateStatusOpts)
+		return nil
+	}
+
+	shouldCountTowardsMax := true
+	failurePolicyRecreateAll(ctx, js, shouldCountTowardsMax, updateStatusOpts)
+	return nil
+}
+
+// restartJobSetAndIgnoreMaxRestartsActionApplier applies the RestartJobSetAndIgnoreMaxRestarts FailurePolicyAction
+var restartJobSetAndIgnoreMaxRestartsActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
+	shouldCountTowardsMax := false
+	failurePolicyRecreateAll(ctx, js, shouldCountTowardsMax, updateStatusOpts)
+	return nil
+}
+
+// applyFailurePolicyRuleAction applies the supplied FailurePolicyRuleAction.
+func applyFailurePolicyRuleAction(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOps *statusUpdateOpts, failurePolicyRuleAction jobset.FailurePolicyAction) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	applier, ok := actionFunctionMap[failurePolicyRuleAction]
+	if !ok {
+		err := fmt.Errorf("failed to find a corresponding action for the FailurePolicyRuleAction %v", failurePolicyRuleAction)
+		log.Error(err, "retrieving information for FailurePolicyRuleAction")
+		return err
+	}
+
+	err := applier(ctx, js, matchingFailedJob, updateStatusOps)
+	if err != nil {
+		log.Error(err, "error applying the FailurePolicyRuleAction: %v", failurePolicyRuleAction)
+		return err
+	}
+
```

Comment:
My sense is that a simple switch statement here is easier to read than defining a map at the top of the file since this is a short list, it would also make the code a little lighter since we would not need to define a function signature:

```
func applyFailurePolicyRuleAction( ... ) {
  err := nil
  switch failurePolicyRuleAction {
    jobset.FailJobSet:
      err = applyFailJobSetAction(...)
    jobset.RestartJobSet:
      err = applyRestartJobSetAction(...)
    jobset.RestartJobSetAndIgnoreMaxRestarts:
      err = applyRestartJobSetAndIgnoreMaxRestartsAction(...)
    default:
      err = fmt.Errorf(...)
  }
  return err
}
```

Reply:
My perspective is that long switch statements are not actually read in detail and clutter the logic of the code. I do not think this is an issue when there are a few number of cases in the switch statement.

My expectation is that we will add more failure policy action types over time and therefore the switch statement would become very long. My intention here was to plan ahead for that so that a larger refactor would not be needed in future.

I also do not think that the particulars of the failure policy action types (such as their names and the call back function which implements their logic) add much to understanding how the function `applyFailurePolicyRuleAction` works.

Reply:
I think I will keep it as is since my expectation is the list will grow over time. Thank you for the feedback.